service/storage_service.cc

```diff
@@ -460,68 +460,92 @@ future<> storage_service::merge_topology_snapshot(raft_topology_snapshot snp) {
+class topology_mutation_builder;
+
+class topology_node_mutation_builder {
     schema_ptr _s;
-    mutation _m;
     api::timestamp_type _ts;
+    mutation& _m;
     deletable_row& _r;
+
 public:
-    topology_mutation_builder(api::timestamp_type ts, raft::server_id);
+    topology_node_mutation_builder(topology_mutation_builder&, raft::server_id);
+
     template<typename T>
-    topology_mutation_builder& set(const char* cell, const T& value) {
+    topology_node_mutation_builder& set(const char* cell, const T& value) {
         return set(cell, sstring{fmt::format("{}", value)});
     }
-    topology_mutation_builder& set(const char* cell, const sstring& value);
-    topology_mutation_builder& set(const char* cell, const raft::server_id& value);
-    topology_mutation_builder& set(const char* cell, const std::unordered_set<dht::token>& value);
-    topology_mutation_builder& set(const char* cell, const uint32_t& value);
-    topology_mutation_builder& set(const char* cell, const utils::UUID& value);
+    topology_node_mutation_builder& set(const char* cell, const sstring& value);
+    topology_node_mutation_builder& set(const char* cell, const raft::server_id& value);
+    topology_node_mutation_builder& set(const char* cell, const std::unordered_set<dht::token>& value);
+    topology_node_mutation_builder& set(const char* cell, const uint32_t& value);
+    topology_node_mutation_builder& set(const char* cell, const utils::UUID& value);
+    topology_node_mutation_builder& del(const char* cell);
+    canonical_mutation build() { return canonical_mutation{std::move(_m)}; }
+};
+
+class topology_mutation_builder {
+    friend class topology_node_mutation_builder;
+
+    schema_ptr _s;
+    mutation _m;
+    api::timestamp_type _ts;
+
+    std::optional<topology_node_mutation_builder> _node_builder;
+public:
+    topology_mutation_builder(api::timestamp_type ts);
     topology_mutation_builder& set_transition_state(topology::transition_state);
     topology_mutation_builder& set_current_cdc_generation_id(const cdc::generation_id_v2&);
     topology_mutation_builder& set_new_cdc_generation_data_uuid(const utils::UUID& value);
     topology_mutation_builder& del_transition_state();
-    topology_mutation_builder& del(const char* cell);
+    topology_node_mutation_builder& with_node(raft::server_id);
     canonical_mutation build() { return canonical_mutation{std::move(_m)}; }
 };
 
-topology_mutation_builder::topology_mutation_builder(api::timestamp_type ts, raft::server_id id) :
+topology_mutation_builder::topology_mutation_builder(api::timestamp_type ts) :
         _s(db::system_keyspace::topology()),
         _m(_s, partition_key::from_singular(*_s, db::system_keyspace::TOPOLOGY)),
-        _ts(ts),
+        _ts(ts) {
+}
+
+topology_node_mutation_builder::topology_node_mutation_builder(topology_mutation_builder& builder, raft::server_id id) :
+        _s(builder._s),
+        _ts(builder._ts),
+        _m(builder._m),
```

Comment:
Isn't it simple to simply store a reference to the `builder`?

Reply:
Ok, I'll see what I can do

Reply:
Ok